### PR TITLE
fix: The chunk files remain in the object storage even if the PVC is …

### PR DESCRIPTION
**Pull request by kyungwan-nam**
…deleted

https://github.com/juicedata/juicefs-csi-driver/issues/738

The job for volume deletion will be terminated immediately after running ``juicefs rmr``.

I added ``juicefs gc --delete``to clean the data in the object storage.

### CODE REVIEW

File: pkg/juicefs/mount/builder/job.go [hunk 1]
```diff
@@ -135,7 +135,7 @@ func (r *Builder) getDeleteVolumeCmd() string {
 	} else {
 		jfsPath = config.CliPath
 	}
-	return fmt.Sprintf("%s && if [ -d /mnt/jfs/%s ]; then %s rmr /mnt/jfs/%s; fi;", cmd, r.jfsSetting.SubPath, jfsPath, r.jfsSetting.SubPath)
+	return fmt.Sprintf("%s && if [ -d /mnt/jfs/%s ]; then %s rmr /mnt/jfs/%s && %s gc ${metaurl} --delete; fi;", cmd, r.jfsSetting.SubPath, jfsPath, r.jfsSetting.SubPath, jfsPath)
```
Comment (zwwhdls) [inline pkg/juicefs/mount/builder/job.go]: `gc` is a little different between [ce](https://juicefs.com/docs/zh/community/command_reference#gc) (`juicefs gc $METAURL --delete`) and [ee](https://juicefs.com/docs/zh/cloud/reference/command_reference#gc) (`juicefs gc $VOL_NAME --delete`).

**Review by zwwhdls — CHANGES_REQUESTED** [review]

File: pkg/juicefs/mount/builder/job.go (hunk 1 above)
Comment (kyungwan-nam) [reply pkg/juicefs/mount/builder/job.go]: I fixed it.
I can't test it under ``ee``, but it works well on ``ce``.

File: pkg/juicefs/mount/builder/job.go [hunk 2]
```diff
@@ -130,12 +130,15 @@ func (r *Builder) getCreateVolumeCmd() string {
 func (r *Builder) getDeleteVolumeCmd() string {
 	cmd := r.getJobCommand()
 	var jfsPath string
+	var gcArgs string
 	if r.jfsSetting.IsCe {
 		jfsPath = config.CeCliPath
+		gcArgs = "${metaurl}"
 	} else {
 		jfsPath = config.CliPath
+		gcArgs = "${VOL_NAME}"
```
Comment (zwwhdls) [inline pkg/juicefs/mount/builder/job.go]: You can use `r.jfsSetting.Name`

**Comment by Hexilee** [comment]
I do not recommend to execute `gc` automatically because of some reasons:

1. `gc` will scan the whole file system, including metadata and object storage, which may be very slow.
2. Even if you executed the `gc` command, the chunk files may still exist in the object storage because JuiceFS enables the trash directory by default.

So, I recommend you to clean up the trash directory as necessary and run `gc` command mannelly.

**Comment by kyungwan-nam** [comment]
@Hexilee 

Thanks for your comment! It makes sense.
I think ``trash`` is not useful on [dynamic-provisioning](https://juicefs.com/docs/csi/guide/pv#dynamic-provisioning)
Users will expect to clean their data when deleting PVC which is ``Delete`` reclaimPolicy. Also, It doesn't need to be recovered. so, the trash is not enabled in our cluster.
It would be good if it is configurable not default. please let me know your thoughts. thanks!

**Comment by Hexilee** [comment]
> @Hexilee
> 
> Thanks for your comment! It makes sense. I think `trash` is not useful on [dynamic-provisioning](https://juicefs.com/docs/csi/guide/pv#dynamic-provisioning) Users will expect to clean their data when deleting PVC which is `Delete` reclaimPolicy. Also, It doesn't need to be recovered. so, the trash is not enabled in our cluster. It would be good if it is configurable not default. please let me know your thoughts. thanks!

Well, if you disabled the trash directory by `juicefs config <metaurl> --trash-days 0`, the chunk objects are also deleted when files are deleted. However, because the objects are deleted asynchronously, you may wait for some time to see the results. 

In this case, you should execute `gc` command only when there are some leaked objects.

**Comment by kyungwan-nam** [comment]
@Hexilee 

> However, because the objects are deleted asynchronously, you may wait for some time to see the results.

Even if I wait, it may not be deleted. because the juices client does not exist after the PVC is deleted.

**Comment by Hexilee** [comment]
> deleted

Other juicefs clients will continue to do it. You can execute `gc` manually If there is no other active PVC in a long time.

**Comment by kyungwan-nam** [comment]
@Hexilee @zwwhdls 
Thank you for letting me know.
I understood this causes bad effects as your comment.
I will try another way to solve it. thanks!

**Comment by Hexilee** [comment]
@kyungwan-nam Thanks and look forward to your contribution!

**Pull request closed without merging**